src/components/Scaler/index.js

```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
+    const newSubDomains = { ...subDomainsByItemId };
+    Object.keys(changedDomainsById).forEach(itemId => {
+      newSubDomains[itemId] = { ...(subDomainsByItemId[itemId] || {}) };
+      Object.keys(changedDomainsById[itemId]).forEach(axis => {
+        let newSubDomain = changedDomainsById[itemId][axis];
+        const newSpan = newSubDomain[1] - newSubDomain[0];
+
+        const existingSubDomain =
+          subDomainsByItemId[itemId][axis] || newSubDomain;
+        const existingSpan = existingSubDomain[1] - existingSubDomain[0];
+
+        const limits = ((domainsById || {})[itemId] || {})[axis] ||
+          (axis === String(Axes.time)
+            ? // FIXME: Phase out this single timeDomain thing.
+              this.props.dataContext.timeDomain
+            : undefined) || [Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER];
 
-    return newXSubDomain;
+        if (newSpan === existingSpan) {
+          // This is a translation; check the bounds.
+          if (newSubDomain[0] <= limits[0]) {
+            newSubDomain = [limits[0], limits[0] + newSpan];
+          }
+          if (newSubDomain[1] >= limits[1]) {
+            newSubDomain = [limits[1] - newSpan, limits[1]];
+          }
+        } else {
+          newSubDomain = [
+            Math.max(limits[0], newSubDomain[0]),
+            Math.min(limits[1], newSubDomain[1]),
+          ];
+        }
+
+        newSubDomains[itemId][axis] = newSubDomain;
+
+        if (axis === String(Axes.time)) {
+          newTimeSubDomain = newSubDomain;
+        }
+      });
+    });
+    this.setState(
+      { subDomainsByItemId: newSubDomains },
+      callback ? () => callback(changedDomainsById) : undefined
+    );
+    if (newTimeSubDomain) {
+      this.props.dataContext.timeSubDomainChanged(newTimeSubDomain);
+    }
   };
 
   render() {
-    const { ySubDomains, yTransformations, xSubDomain } = this.state;
+    const { domainsByItemId, subDomainsByItemId, timeSubDomain } = this.state;
     const { dataContext, xScalerFactory } = this.props;
     const ownContext = {
-      updateXTransformation: this.updateXTransformation,
-      updateXSubDomain: this.updateXSubDomain,
-      updateYTransformation: this.updateYTransformation,
-      yTransformations,
+      updateDomains: this.updateDomains,
+      domainsByItemId,
+      subDomainsByItemId,
     };
 
-    const enrichedSeries = dataContext.series.map(s => ({
-      ...s,
-      ySubDomain: ySubDomains[s.id] || s.ySubDomain,
-    }));
-
-    const enrichedCollections = dataContext.collections.map(c => ({
-      ...c,
-      ySubDomain: ySubDomains[c.id] || c.ySubDomain,
-    }));
-
     const enrichedContext = {
-      collections: enrichedCollections,
-      series: enrichedSeries,
-      xSubDomain: xSubDomain || dataContext.xSubDomain,
+      collections: dataContext.collections,
+      series: dataContext.series,
+      timeSubDomain: timeSubDomain || dataContext.timeSubDomain,
```

Comment:
I don't actually see the `timeSubDomain` state being used anywhere in Scaler other than here. In which case, can we just remove it from the state and have `dataContext.timeSubDomain || dataContext.timeDomain` here?

Reply:
Amazing find! Yep, there's no need to have it here.